lib/internal/quic/core.js

```diff
@@ -2590,11 +2651,64 @@ class QuicStream extends Duplex {
   }
 
   [kStreamReset](code) {
-    this[kInternalState].resetCode = code | 0;
+    // Receiving a reset from the peer indicates that it is no
+    // longer sending any data, we can safely close the readable
+    // side of the Duplex here.
+    this[kInternalState].resetCode = code;
     this.push(null);
     this.read();
   }
 
+  [kClose]() {
+    const state = this[kInternalState];
+
+    if (this.destroyed) {
+      return PromiseReject(
+        new ERR_INVALID_STATE('QuicStream is already destroyed'));
+    }
+
+    const promise = deferredClosePromise(state);
+    if (this.readable) {
+      this.push(null);
+      this.read();
+    }
+
+    if (this.writable) {
+      this.end();
+    }
+
+    return promise;
+  }
+
+  close() {
+    return this[kInternalState].closePromise || this[kClose]();
+  }
+
+  _destroy(error, callback) {
+    const state = this[kInternalState];
+    const handle = this[kHandle];
+    this[kSetHandle]();
+    if (handle !== undefined)
+      handle.destroy();
+    state.session[kRemoveStream](this);
+
+    if (error && typeof state.closePromiseReject === 'function')
+      state.closePromiseReject(error);
+    else if (typeof state.closePromiseResolve === 'function')
+      state.closePromiseResolve();
+
+    process.nextTick(() => callback(error));
```

Comment:
Why nextTick here?

Reply:
Because previously it wasn't working correctly without it. I haven't tested that since.

Reply:
Yeah, quick test shows that there are some timing issues on cleanup when this is not deferred to nextTick. We can explore addressing those later to see if we can remove the nextTick in the future but for now, it's needed here.

Reply:
Could you please add a TODO comment here so we can revisit it later?